src/backer/crawling.py

```diff
@@ -32,14 +33,25 @@ def crawlBlockDo(self, tymth=None, tock=0.0):
             _, _, _ = self.client.find_intersection.execute([tip.to_point()])
 
             while True:
-                direction, tip, block, _ = self.client.next_block.execute()
+                direction, tip, block, _ = self.client.next_block.execute()                
 
                 if direction == ogmios.Direction.forward:
+                    if tip.height:
+                        self.ledger.updateTip(tip.height)
+
                     if not self.on_tip and block.height == tip.height:
                         logger.info(f"Reached tip at slot {block.slot}")
                         self.on_tip = True
                         self.extend(self.backer)
                         self.tock = 1.0
+                    
+                    # Find transactions involving cardano backer
+                    if self.on_tip and isinstance(block, ogmios.Block) and hasattr(block, "transactions"):
+                        for tx in block.transactions:                            
+                            txId = tx['inputs'][0]['transaction']['id']                            
+                            
+                            if self.ledger.getConfirmingTrans(txId) is not None:
+                                self.ledger.updateTrans(txId, block.height, block.slot)
```

Comment:
OK, so you are storing the block height and time when it appears, and then comparing it on the side. Cool - but in this case you aren't handling rollbacks.

We need to handle the rollback else block to reduce the tip height and to revert this height and slot set for the txid back to default.

Otherwise if the chain rollbacks, the data will be wrong

Reply:
Can you explain more about "chain rollbacks"? How can we detect it or what field in the transaction of block can express that It is a "rooll back"?

Reply:
A blockchain is decentralised so each node can have a slightly different view of part of the chain (near the tip), but they are eventually consistent. So the deeper a block is in the chain (e.g. 16 is pretty secure) the more likely it is that it will eventually be in every node's view of the chain. After 2160 blocks it's 100% certain, but the probably drops very quickly so 16 is enough.

What this means: if the node we're connected to had a DIFFERENT tip than the eventually consistent one, it will roll back using the consensus mechanism and join the fork that is eventually consistent.

This is where in ogmios the block you receive will be in the direction of BACKWARD, this means it's rolling back. There are often mini rollbacks of 1-2 blocks.

Reply:
So if we update a tx from forward, and then rollback and notice the tx there again, we need to undo the action we took on the tx.